src/main/java/org/commcare/formplayer/beans/menus/EntityListResponse.java

```diff
@@ -73,7 +75,12 @@ public EntityListResponse(EntityScreen nextScreen,
             }
             entities = processEntitiesForCaseDetail(detail, reference, ec, neededDatum);
         } else {
-            Vector<TreeReference> references = nextScreen.getReferences();
+            Vector<TreeReference> references = new Vector<TreeReference>();
+            this.requireSearch = nextScreen.getShortDetail().doesRequireSearch();
+            this.searchText = searchText;
+            if (!this.requireSearch || this.searchText != null && !"".equals(this.searchText)) {
+                references = nextScreen.getReferences();
```

Comment:
Might be worth talking with @calellowitz  about his recent changes.

I believe this current structure will still end up doing a huge amount of the work of the work, because the screen itself will have already generated the reference set before it's requested here. You likely want to bubble up the logic for where the search requirement comes in into the screen itself (which will let you suppress the majority of the work of filtering the reference set and loading fields from I/O)